src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -551,7 +567,15 @@ private ThreadId TryEnterSlow(int timeoutMs, ThreadId currentThreadId)
             return new ThreadId(0);
         }
 
-        private void ResetWaiterStartTime() => _waiterStartTimeMs = 0;
+        private bool UseAlertableWaits => (_waiterStartTimeMsAndFlags & 1) == 0;
+
+        private ushort WaiterStartTimeMs
+        {
+            get => (ushort)(_waiterStartTimeMsAndFlags >> 1);
```

Comment:
`ushort` range in milliseconds is ~65 seconds. This reduces it by half. On the other hand we get these timings from ticks that are typically more coarse than 1 msec.

Overflows here are not dangerous per say and 30 seconds is still a long time, but perhaps it is still better to sacrifice precision rather than the range?
I.E - store the value without shifting, just mask out the lowest bit when returning.

Reply:
I don't think the range matters much, the feature that relies on it operates on 100-ms granularity, which can be served just as well before and after this change.

> I.E - store the value without shifting, just mask out the lowest bit when returning.

I didn't follow, can you elaborate?

Reply:
I was thinking of something like the following:

```cs
        private ushort WaiterStartTimeMs
        {
            get => (ushort)(_waiterStartTimeMsAndFlags | 1);
            set => _waiterStartTimeMsAndFlags = (ushort)(value | (_waiterStartTimeMsAndFlags & 1));
        }
```

In fact the `| 1` in the getter might not be necessary, considering ticks are coarser than 1msec.

It is ok either way though.

Reply:
What's the benefit of that?

Reply:
> What's the benefit of that?

Smaller chance of wrapping around if something gets stuck for 30 seconds. When it is a matter of seconds, things can realistically happen.

It is a very small benefit though since it would be rare and wrap around is not a huge deal.

Reply:
I see, I don't think it's necessary to do something like that at the moment, there isn't an issue with the range.